### PR TITLE
Queuing

### LINKED ISSUE #14: Move to PostgreSQL database

### LINKED ISSUE #15: Use queueing for GitHub webhook

### PR DESCRIPTION
**Pull request by Tioborto**
### Changed

* Use Celery instead of background tasks for Webhook analysis
* Use async database

### Added

* Pre-commits
* PgAdmin
* Flower

##### ISSUE TYPE

- Feature Pull Request

### CODE REVIEW

File: pylintrc [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        logging-format-interpolation, # we use f-strings, it works great and isn't slower
+        too-few-public-methods, # this conflicts with Pydantic classes, which are really dataclasses without any methods (usually)
+        bad-continuation, # we use black for formatting, which interferes with Pylint rules
+        fixme, # Leaving TODOs in code is OK
+
+[IMPORTS]
+ignored-modules=
+
+[TYPECHECK]
+ignored-classes=responses # Responses defines all its global methods dynamically, which is something Pylint doesn't understand
```
Comment (nymous) [inline pylintrc]: Hehe you copied a bit too much from the example file ^^
You don't use [getsentry/responses](https://github.com/getsentry/responses) yet, and they even fixed this issue in version [0.10.7](https://github.com/getsentry/responses/releases/tag/0.10.7)

File: raddar/api/api_v1/endpoints/projects.py [hunk 2]
```diff
@@ -1,32 +1,16 @@
...
 
-@router.get(
-    "/{project_name:path}/refs/{ref_name:path}", response_model=schemas.Analysis
-)
-def get_project_head_secrets(
-    project_name: str, ref_name: str, db: Session = Depends(dependencies.get_db)
-):
-    return crud.get_analysis_by_name_and_ref(
-        db=db, project_name=project_name, branch_name=ref_name, ref_name=ref_name
-    )
```
Comment (nymous) [inline raddar/api/api_v1/endpoints/projects.py]: Woah you removed this whole endpoint?!

File: raddar/api/api_v1/endpoints/webhooks.py [hunk 3]
```diff
@@ -1,27 +1,18 @@
...
-        analysis=schemas.AnalysisBase(branch_name=get_branch_name(payload.ref)),
-        scan_origin="github-webhook",
-        db=db,
+def scan_github_project(payload: models.GitHubPushPayload):
+    analysis = schemas.AnalysisBase(branch_name=get_branch_name(payload.ref))
+    background_id = background_project_analysis.delay(
+        payload.repository.full_name, analysis.dict(), "github-webhook"
     )
-    return {"message": "Notification sent in the background"}
+    return {"message": f"Notification sent in the background : {background_id}"}
```
Comment (nymous) [inline raddar/api/api_v1/endpoints/webhooks.py]: Maybe a message like "Scan scheduled in the background with id {background_id}" would better convey what is happening.
Also sending this id as a field on its own could be interesting, if the caller wants to get the scan status later (one the endpoint is implemented of course), so adding something like `"background_id": background_id` or `"scan_id": background_id`?

File: raddar/core/celery_app.py [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    print("Initializing database connection for worker.")
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(database.connect())
+
+
+@worker_process_shutdown.connect
+def shutdown(**kwargs):
+    print("Closing database connection for worker.")
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(database.disconnect())
```
Comment (nymous) [inline raddar/core/celery_app.py:21]: YEEEES, we found how to do it!!! It was painful but worth it!!
(You can resolve this message ^^)

File: raddar/core/celery_app.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+
+celery_app = Celery(
+    "raddar",
+    broker="pyamqp://guest@queue//",
```
Comment (nymous) [inline raddar/core/celery_app.py]: Maybe this URI could come from a settings value, so it can be configured to point to any RabbitMQ deployed by the user?

File: raddar/core/settings.py [hunk 6]
```diff
@@ -1,12 +1,37 @@
...
+    ) -> Any:
+        if isinstance(value, str):
+            return value
+        return PostgresDsn.build(
+            scheme="postgresql",
+            user=values.get("POSTGRES_USER"),
+            password=values.get("POSTGRES_PASSWORD").get_secret_value(),
+            host=values.get("POSTGRES_SERVER"),
+            path=f"/{values.get('POSTGRES_DB') or ''}",
+        )
```
Comment (nymous) [inline raddar/core/settings.py]: Wow, where did you find the reference for this?

File: raddar/crud/crud.py [hunk 7]
```diff
@@ -1,68 +1,66 @@
...
-def create_analysis_secret(db: Session, secret: schemas.SecretCreate, analysis_id: int):
-    db_secret = models.Secret(**secret, analysis_id=analysis_id)
-    db.add(db_secret)
-    db.commit()
-    db.refresh(db_secret)
-    return db_secret
+async def create_analysis_secret(
+    secret_to_create: schemas.SecretBase, analysis_id: int
+):
+    query = secret.insert(None).values(
```
Comment (nymous) [inline raddar/crud/crud.py:15]: Do you have to pass `None` here? The FastAPI docs don't pass any parameters there.

File: raddar/crud/crud.py [hunk 8]
```diff
@@ -1,68 +1,66 @@
...
 ):
-    db_project = get_project_by_name(db, project_name=project.name)
-
-    if not db_project:
-        db_project = create_project(db, project)
-
-    db_analysis = models.Analysis(
-        execution_date=datetime.now(),
-        branch_name=branch_name,
+    now = datetime.now()
```
Comment (nymous) [inline raddar/crud/crud.py]: Be careful, this can be something else than UTC, and it's not timezone aware (it's a "naive" datetime). You could prefer using `datetime.now(timzeone.utc)` (with a `from datetime import timezone`). See the Python [datetime docs](https://docs.python.org/3/library/datetime.html#datetime.datetime.now) and especially the red warning box of the `utcnow()` method for more information.

File: raddar/db/database.py [hunk 9]
```diff
@@ -1,12 +1,41 @@
...
+from raddar.core.settings import settings
 
-engine = create_engine(
-    SQLALCHEMY_DATABASE_URL, connect_args={"check_same_thread": False}
+metadata = sqlalchemy.MetaData()
+
+project = Table(
+    "project",
+    metadata,
+    Column("id", Integer, primary_key=True),
```
Comment (nymous) [inline raddar/db/database.py:12]: You skipped the `unique=True` and `index=True` that were on the model previously, is this expected?

File: raddar/db/database.py [hunk 10]
```diff
@@ -1,12 +1,41 @@
...
+    Column("name", String, unique=True),
```
Comment (nymous) [inline raddar/db/database.py]: Same here, you replaced the `index=True` with a `unique=True`. I think a UNIQUE column has to be indexed so it should be ok.

File: raddar/db/database.py [hunk 11]
```diff
@@ -1,12 +1,41 @@
...
-Base = declarative_base()
+secret = Table(
+    "secret",
+    metadata,
+    Column("id", Integer, primary_key=True),
+    Column("filename", String),
+    Column("secret_type", String),
+    Column("line_number", Integer),
+    Column("secret_hashed", String),
+    Column("analysis_id", Integer, ForeignKey("analysis.id"), nullable=False),
```
Comment (nymous) [inline raddar/db/database.py]: On the other hand, good thing you didn't keep the index on *all* columns like before :D

File: raddar/lib/custom_typing.py [hunk 12]
```diff
@@ -0,0 +1,3 @@
+from typing import Literal
+
+ScanOrigin = Literal["manual", "github-webhook"]
```
Comment (nymous) [inline raddar/lib/custom_typing.py]: Not sure about this being in its own file, I would see it more with the other models

File: raddar/lib/managers/detect_secrets_manager.py [hunk 13]
```diff
@@ -1,44 +1,71 @@
...
         for file in baseline["results"]:
             for secret in baseline["results"][file]:
-                new_secret = {}
-                new_secret["filename"] = file.split(f"{temp_dir}/")[1]
-                new_secret["secret_type"] = secret["type"]
-                new_secret["line_number"] = secret["line_number"]
-                new_secret["secret_hashed"] = secret["hashed_secret"]
-                crud.create_analysis_secret(db, new_secret, analysis_returned.id)
+                new_secret = schemas.SecretBase(
+                    filename=file.split(f"{temp_dir}/")[1],
```
Comment (nymous) [inline raddar/lib/managers/detect_secrets_manager.py]: I know this would probably never happen, but adding a `maxsplit=1` parameter here would make sure we only ever have two parts.

File: raddar/lib/managers/repository_manager.py [hunk 14]
```diff
@@ -1,29 +1,22 @@
...
-    FailedToCloneRepoException,
-    FailedToWriteRepoException,
-)
+from ..exception import FailedToCloneRepoException, FailedToWriteRepoException
 
 
 def get_branch_name(ref_name: str) -> str:
     if ref_name.startswith("refs/"):
-        ref_name = re.sub("^refs\/\\w+\/", "", ref_name, count=1)
+        ref_name = re.sub(r"^refs/\w+/", "", ref_name, count=1)
```
Comment (nymous) [inline raddar/lib/managers/repository_manager.py:10]: Nice, way better like this :+1:

File: raddar/schemas/schemas.py [hunk 15]
```diff
@@ -18,12 +18,9 @@ class SecretCreate(SecretBase):
 class Secret(SecretBase):
     id: int
 
-    class Config:
-        orm_mode = True
```
Comment (nymous) [inline raddar/schemas/schemas.py]: Why did you remove the `orm_mode` from the models?

File: raddar/main.py [hunk 16]
```diff
@@ -1,16 +1,21 @@
...
 
 from raddar.api.api_v1.api import api_router
-from raddar.db.database import engine
-from raddar.models import models
 from raddar.core.settings import settings
+from raddar.db.database import database
 
+app = FastAPI(
+    title=settings.PROJECT_NAME, openapi_url=f"{settings.API_V1_STR}/openapi.json"
+)
```
Comment (nymous) [inline raddar/main.py]: Why do you specify the `openapi_url` here?
If you ever do a v2 (attached to another APIRouter) it will be included in the same openapi.json, so you will have a `/api/v1/openapi.json` with both v1 and v2 endpoints...

**Review by nymous — CHANGES_REQUESTED** [review]

File: raddar/api/api_v1/endpoints/webhooks.py (hunk 3 above)
Comment (Tioborto) [reply raddar/api/api_v1/endpoints/webhooks.py]: I will return a message and the scan_id it's a good id:
```
return {
        "message": "Scan scheduled in the background",
        "scan_id": str(background_id),
 }
```

File: raddar/core/settings.py (hunk 6 above)
Comment (Tioborto) [reply raddar/core/settings.py]: In the cloud 😂

Comment (Tioborto) [reply raddar/core/settings.py]: https://pydantic-docs.helpmanual.io/usage/validators/

File: raddar/crud/crud.py (hunk 7 above)
Comment (Tioborto) [reply raddar/crud/crud.py:13]: Removed

Comment (Tioborto) [reply raddar/crud/crud.py:15]: Pylint send an error when None is removed. 
This is the header of this function 'insert' : 
`def insert(self, dml, values=None, inline=False, **kwargs):`
A dml is expected..

File: raddar/crud/crud.py (hunk 8 above)
Comment (Tioborto) [reply raddar/crud/crud.py]: Great to know 👍

File: raddar/db/database.py (hunk 9 above)
Comment (Tioborto) [reply raddar/db/database.py:12]: a primary key is unique by default i mean

File: raddar/db/database.py (hunk 10 above)
Comment (Tioborto) [reply raddar/db/database.py]: I forget the index arg. I added it

File: raddar/lib/managers/detect_secrets_manager.py (hunk 13 above)
Comment (Tioborto) [reply raddar/lib/managers/detect_secrets_manager.py]: well seen

File: raddar/lib/custom_typing.py (hunk 12 above)
Comment (Tioborto) [reply raddar/lib/custom_typing.py]: File removed and definition moved to models/models.py

File: raddar/db/database.py (hunk 11 above)
Comment (Tioborto) [reply raddar/db/database.py]: I added some `index=True`

File: raddar/api/api_v1/endpoints/projects.py (hunk 2 above)
Comment (Tioborto) [reply raddar/api/api_v1/endpoints/projects.py]: Added

File: raddar/core/celery_app.py (hunk 5 above)
Comment (Tioborto) [reply raddar/core/celery_app.py]: Added in .env_template file

File: raddar/schemas/schemas.py (hunk 15 above)
Comment (Tioborto) [reply raddar/schemas/schemas.py]: I thought it was no longer useful, but in fact it allows you to retrieve objects and access them by doing `item.id`

File: pyproject.toml [hunk 17]
```diff
@@ -11,13 +11,19 @@ uvicorn = "^0.12.2"
 detect-secrets = "^0.14.3"
 SQLAlchemy = "^1.3.20"
 GitPython = "^3.1.9"
-pydantic = {extras = ["dotenv"], version = "^1.6.1"}
+pydantic = {extras = ["typing_extensions"], version = "^1.7.3"}
```
Comment (nymous) [inline pyproject.toml]: Be careful you lost the `dotenv` extension here. And do you need `typing_extensions`? The Pydantic docs say "use of Literal prior to python 3.8 relies on typing-extensions" and you're on 3.8+

File: raddar/crud/crud.py [hunk 18]
```diff
@@ -1,68 +1,80 @@
...
-        .filter(
-            models.Project.name == project_name,
-            or_(ref_name == ref_name, branch_name == branch_name),
+    query = project.select().where(
+        and_(
+            project.c.name == project_name,
+            or_(analysis.c.branch_name == branch_name, analysis.c.ref_name == ref_name),
         )
-        .order_by(models.Analysis.execution_date.desc())
-        .first()
```
Comment (nymous) [inline raddar/crud/crud.py:67]: You removed the `order_by` and the `first`?

**Review by nymous — CHANGES_REQUESTED** [review]
Last changes needed and it's good to go!

File: pyproject.toml (hunk 17 above)
Comment (Tioborto) [reply pyproject.toml]: When I do poetry add pydantic[dotenv], it will remove typing_extensions and not append dotenv package
I added dotenv and removed typing_extensions

Comment (nymous) [reply pyproject.toml]: Yes, if you need multiple extras you have to specify them all in one go, like `poetry add requests[security,socks]` (https://python-poetry.org/docs/cli/#add)

File: raddar/crud/crud.py (hunk 18 above)
Comment (Tioborto) [reply raddar/crud/crud.py:67]: refactor ^^

**Review by nymous — APPROVED** [review]
Aaaaaand it's good!
I'm not a big fan of this construct
```py
    secret_list = []
    for secret in project_analysis_returned:
        secret_list.append(dict(secret))
```
but I'm not sure the alternative is better (or if it even works):
```py
    # This should work
    secret_list = [dict(secret) for secret in project_analysis_returned]
    # This I'm not sure, maybe it doesn't like the generator. But if it does it would be great, especially if there are many many secrets, because generators are more memory efficient
    secret_list = (dict(secret) for secret in project_analysis_returned)
    return secret_list
```
